src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -0,0 +1,619 @@
+                if (aritiesSeenSoFar.ContainsKey(arity))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.ClassConstructorRule, constructor.GetLocation(), classDeclaration.Identifier, arity));
+                    return true;
+                }
+                else
+                {
+                    aritiesSeenSoFar[arity] = true;
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Checks each type in the given list of types and sees if any are equal to the given type name
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="context"></param>
+        /// <param name="maybeQualName">A list of the descendent nodes that are of the given type, possibly empty. 
+        /// empty example: this property doesnt have any qualified types in its signature
+        /// </param>
+        /// <param name="typeName">check to see if this type appears in the signature</param>
+        /// <param name="classId">name of the class this field is in</param>
+        /// <param name="loc">syntax location</param>
+        /// <param name="fieldKind">either property or method</param>
+        /// <returns>true if the given type is the same as the one in the list</returns>
+        private bool SignatureContainsTypeName<T>(ref GeneratorExecutionContext context, 
+            IEnumerable<T> typesInSignature, 
+            string typeName, 
+            SyntaxToken classId, 
+            Location loc, 
+            SyntaxToken fieldKind)
+        {
+            foreach (T name in typesInSignature)
+            {
+                if (name.ToString().Equals(typeName))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.ArraySignature_SystemArrayRule, loc, classId, fieldKind)); 
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Checks to see if the class declares any operators (overloading them)
+        /// </summary>
+        /// <param name="context">Compilation unit</param>
+        /// <param name="classDeclaration">
+        /// Class to check for operator declarations 
+        /// operator declarations are just like method declarations except they use the `operator` keyword 
+        /// </param>
+        /// <returns>True iff an operator is overloaded by the given class</returns>
+        public bool OverloadsOperator(ref GeneratorExecutionContext context, ClassDeclarationSyntax classDeclaration)
+        {
+            var operatorDeclarations = classDeclaration.DescendantNodes().OfType<OperatorDeclarationSyntax>();
+            foreach (var op in operatorDeclarations)
+            { 
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.OperatorOverloadedRule, op.GetLocation(), op.OperatorToken));

```

Comment:
We should probably remove the return true and then change return false to return found or return operatorDeclarations.Count != 0

Reply:
I'll try this out